src/state-item.c

```diff
@@ -0,0 +1,417 @@
+//
+
+#include <config.h>
+#include <stdlib.h>
+#include "state-item.h"
+#include "closure.h"
+#include "nullable.h"
+#include <time.h>
+#include <gl_linked_list.h>
+#include <gl_xlist.h>
+#include <assert.h>
+#include "getargs.h"
+
+size_t nstate_items;
+state_item_number *state_item_map;
+state_item *state_items;
+
+state_item_number *trans;
+bitsetv rev_trans;
+Hash_table *prods;
+Hash_table *rev_prods;
+
+
+// hash functions for index -> bitset hash maps
+typedef struct
+{
+    int key;
+    bitset l;
+} hash_pair;
+
+static size_t
+hash_pair_hasher(const void *sl, size_t max)
+{
+    int key = ((hash_pair*)sl)->key;
+    return key % max;
+}
+
+static bool
+hash_pair_comparator(const void *l, const void *r)
+{
+    return ((hash_pair *)l)->key == ((hash_pair *)r)->key;
+}
+
+static bitset
+hash_pair_lookup(Hash_table *tab, int key)
+{
+    hash_pair *l = xmalloc(sizeof(hash_pair));
+    l->key = key;
+    hash_pair *hp = (hash_pair *)hash_lookup(tab, l);
+    if (hp == NULL) return NULL;
+    return hp->l;
+}
+
+static void
+hash_pair_insert(Hash_table *tab, int key, bitset val)
+{
+    hash_pair *hp = xmalloc(sizeof(hash_pair));
+    hp->key = key;
+    hp->l = val;
+    if (!hash_insert(tab, hp))
+        xalloc_die();
+}
+
+/* return a state_item from a state's id and the offset of the item
+   within the state.
+ */
+state_item*
+state_item_lookup(state_number s, state_item_number off)
+{
+    return &state_items[state_item_index_lookup(s, off)];
+}
+
+static inline void
+state_item_set(state_item_number sidx, state *s, item_number off)
+{
+    state_item *si = state_items + sidx;
+    si->state = s;
+    si->item = &ritem[off];
+    si->reduction_offset = -1;
+    si->num = sidx;
+    trans[sidx] = -1;
+}
+/**
+ * Initialize state_items set
+ */
+static void
+init_state_items()
+{
+    nstate_items = 0;
+    bitsetv production_items = bitsetv_create(nstates, nritems, BITSET_SPARSE);
+    for (int i = 0; i < nstates; i++)
+    {
+        state *s = states[i];
+        nstate_items += s->nitems;
+        closure (s->items, s->nitems);
+        for (size_t j = 0; j < nitemset; j++)
+            if (itemset[j] > 0 && item_number_is_rule_number(ritem[itemset[j] - 1]))
+            {
+                bitset_set(production_items[i], itemset[j]);
+                nstate_items++;
+            }
+    }
+    state_item_map = xnmalloc(nstates + 1, sizeof(state_item_number));
+    state_items = xnmalloc(nstate_items, sizeof(state_item));
+    trans = xnmalloc(nstate_items, sizeof(state_item_number));
+    rev_trans = bitsetv_create(nstate_items, nstate_items, BITSET_SPARSE);
+    state_item_number sidx = 0;
+    for (int i = 0; i < nstates; i++)
+    {
+        state_item_map[i] = sidx;
+        int rule_search_idx = 0;
+        state *s = states[i];
+        reductions *red = s->reductions;
+        for (int j = 0; j < s->nitems; j++)
+        {
+            state_item_set(sidx, s, s->items[j]);
+            state_item *si = state_items + sidx;
```

Comment:
The GNU style here is

``` C
        for (int j = 0; j < s->nitems; j++)
           {
             state_item_set(sidx, s, s->items[j]);
```

You should really look at the rest of Bison, and match the style we follow.  Again, I believe that using clang-format on your files is the safest, simplest, fastest means to do it, without wasting your time.